test/NuGet.Core.Tests/NuGet.Configuration.Test/SearchTreeTest.cs

```diff
@@ -0,0 +1,197 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using FluentAssertions;
+using NuGet.Test.Utility;
+using Xunit;
+
+namespace NuGet.Configuration.Test
+{
+    public class SearchTreeTest
+    {
+        [Theory]
+        [InlineData("public, Nuget", "Nuget")]
+        [InlineData("public, nuget", "Nuget")]
+        [InlineData("public, Nuget", "nuget")]
+        [InlineData("public, nuget", "nuget")]
+        [InlineData("public, Contoso.Opensource.*", "Contoso.Opensource.")]
+        [InlineData("public, Contoso.Opensource.*", "Contoso.Opensource.MVC")]
+        [InlineData("public, Contoso.Opensource.*", "Contoso.Opensource.MVC.ASP")]
+        public void SearchTree_WithOneSource_Match(string packageNamespaces, string term)
+        {
+            // Arrange
+            PackageNamespacesConfiguration configuration = GetPackageNamespacesConfiguration(packageNamespaces);
+            SearchTree searchTree = new SearchTree(configuration);
+
+            // Act & Assert
+            configuration.AreNamespacesEnabled.Should().BeTrue();
+            IReadOnlyList<string> configuredSources = searchTree.GetConfiguredPackageSources(term);
+            Assert.Equal(1, configuredSources.Count);
+            Assert.Equal(configuration.Namespaces.Keys.First(), configuredSources[0]);
+        }
+
+        [Theory]
+        [InlineData("public, nuget", "nuge")]
+        [InlineData("public, nuget", "nuget1")]
+        [InlineData("public, Contoso.Opensource.*", "Cont")]
+        [InlineData("public, Contoso.Opensource.*", "Contoso.Opensource")]
+        public void SearchTree_WithOneSource_NoMatch(string packageNamespaces, string term)
+        {
+            // Arrange
+            SearchTree searchTree = GetSearchTree(packageNamespaces);
+
+            // Act & Assert
+            IReadOnlyList<string> configuredSources = searchTree.GetConfiguredPackageSources(term);
+            Assert.Null(configuredSources);
+        }
+
+        [Theory]
+        [InlineData("nuget.org, nuget | privateRepository, private*", "nuget")]
+        [InlineData("nuget.org, nuget |  privateRepository, private*", "private.")]
```

Comment:
Above line represents below package namespace section and `private.` is the package id.

    <packageNamespaces>
        <packageSource key=""nuget.org"">
            <namespace id=""nuget"" />
        </packageSource>
        <packageSource key=""privateRepository"">
            <namespace id=""private*"" />
        </packageSource>
    </packageNamespaces>

Reply:
Testing insignificant whitespace is important because a) customers may have it and b) the product has code which handles insignificant whitespace.

Here are some possible variations that should be tested.  (Note the insignificant whitespace inside attribute values.)
```XML
<packageNamespaces>
    <packageSource key=" nuget.org ">
        <namespace id=" nuget " />
    </packageSource>
    <packageSource key=" privateRepository ">
        <namespace id=" private* " />
    </packageSource>
</packageNamespaces>
```
Additionally, it is worth testing scenarios where values are absent or an empty string.  Example:
```XML
<packageNamespaces>
    <packageSource key="a.test">
        <namespace id="" />
    </packageSource>
    <packageSource key="a.test">
    </packageSource>
    <packageSource key="">
    </packageSource>
    <packageSource>
    </packageSource>
</packageNamespaces>
```
Some of these are error cases, in which you'd want to verify that the proper exception is thrown.  Many of these scenarios just cover insignificant whitespace trimming.

I'd expect all of these scenarios to be covered somewhere.

As for `SearchTree` tests, I love that you switched to data-driven, Theory-attributed tests.  Great job!  It is much easier than before to reason on what scenarios are covered by existing tests.  The problem is that your inconsistent whitespacing in test data is misleading here and [here](https://github.com/NuGet/NuGet.Client/pull/4039/files#diff-25d12844be3da797979fe4495f9a728c80543a45e0c7cb0023b3e403aa94fb33R66-R68) because the whitespace differences _look_ intentional, yet they are insignificant.  Your [`GetPackageNamespacesConfiguration(...)`](https://github.com/NuGet/NuGet.Client/pull/4039/files#diff-25d12844be3da797979fe4495f9a728c80543a45e0c7cb0023b3e403aa94fb33R167) test method parses and trims these strings, _before testing the product's handling of this input._  The point of this these tests is not to test your test method.

So the whitespace inconsistencies around `|` and `,` here:
```C#
        [InlineData("nuget.org, nuget |  privateRepository, private*", "private.")]
```
...and [here](https://github.com/NuGet/NuGet.Client/pull/4039/files#diff-25d12844be3da797979fe4495f9a728c80543a45e0c7cb0023b3e403aa94fb33R66-R68):
```C#
        [InlineData("nuget.org,nuget | privateRepository,private*", "nuge")]
        [InlineData("nuget.org,nuget | privateRepository,private*", "nuget1")]
        [InlineData("nuget.org,nuget |  privateRepository,private*", "privat")]
```
...are good, but you need to verify that the product handles it correctly and not have [your test helper method prematurely trimming whitespace [here](https://github.com/NuGet/NuGet.Client/pull/4039/files#diff-25d12844be3da797979fe4495f9a728c80543a45e0c7cb0023b3e403aa94fb33R175) and [here](https://github.com/NuGet/NuGet.Client/pull/4039/files#diff-25d12844be3da797979fe4495f9a728c80543a45e0c7cb0023b3e403aa94fb33R186) _before_ the product has a chance to trim.  Otherwise, you're not testing the product.

Here are some simple examples.  (Focus on whitespace, not the text, which may be wrong.)
```C#
        [InlineData("nuget.org,nuget|privateRepository,private*", "term")]
        [InlineData(" nuget.org , nuget | privateRepository , private* ", "term")]
        [InlineData("nuget.org,|privateRepository,private*", "term")]
        [InlineData("nuget.org,|privateRepository,", "term")]
```

Reply:
I refactored the tests. I intention was use whitespaces to make string input more readable. 
Now I removed white space trimmer from my helper and only include white spaces for specific test inputs.
Thank you.